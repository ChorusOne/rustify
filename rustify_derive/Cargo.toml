[package]
name = "rustify_derive"
version = "0.1.0"
edition = "2018"

[lib]
proc-macro = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
syn = "1.0"
quote = "1.0"
synstructure = "0.12.5"
proc-macro2 = "1.0.28"
regex = "1.5.4"
