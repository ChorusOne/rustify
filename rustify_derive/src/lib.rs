//! Provides a derive macro for easily implementing an Endpoint from the
//! `rustify` crate. See the documentation for `rustify` for details on how
//! to use this macro.

#[macro_use]
extern crate synstructure;
extern crate proc_macro;

mod error;
mod params;
mod parse;

use std::collections::{HashMap, HashSet};

use error::Error;
use params::Parameters;
use proc_macro2::Span;
use quote::quote;
use regex::Regex;
use syn::{self, Generics, Ident, Meta, Type};

const MACRO_NAME: &str = "Endpoint";
const ATTR_NAME: &str = "endpoint";
const QUERY_NAME: &str = "query";

fn action(path: &syn::LitStr) -> Result<proc_macro2::TokenStream, Error> {
    let re = Regex::new(r"\{(.*?)\}").unwrap();
    let mut fmt_args: Vec<syn::Expr> = Vec::new();
    for cap in re.captures_iter(path.value().as_str()) {
        let expr = syn::parse_str(&cap[1]);
        match expr {
            Ok(ex) => fmt_args.push(ex),
            Err(_) => {
                return Err(Error::new(
                    path.span(),
                    format!("Failed parsing format argument as expression: {}", &cap[1]).as_str(),
                ));
            }
        }
    }
    let path = syn::LitStr::new(
        re.replace_all(path.value().as_str(), "{}")
            .to_string()
            .as_str(),
        Span::call_site(),
    );

    if !fmt_args.is_empty() {
        Ok(quote! {
            format!(#path, #(#fmt_args),*)
        })
    } else {
        Ok(quote! {
            String::from(#path)
        })
    }
}

fn query(fields: &HashMap<Ident, HashSet<Meta>>) -> proc_macro2::TokenStream {
    // Collect all fields that have a query parameter attached
    let mut query_fields = Vec::<&Ident>::new();
    for (key, value) in fields.iter() {
        for attr in value.iter() {
            if attr.path().is_ident(QUERY_NAME) {
                query_fields.push(key);
            }
        }
    }

    match query_fields.is_empty() {
        false => {
            // Vec of ("#id".to_string(), serde_json::value::to_value(&self.#id).unwrap())
            let exprs = query_fields
                .iter()
                .map(|id| {
                    let id_str = id.to_string();
                    quote! {(#id_str.to_string(), serde_json::value::to_value(&self.#id).unwrap()) }
                })
                .collect::<Vec<proc_macro2::TokenStream>>();

            // Construct query function
            quote! {
                fn query(&self) -> Vec<(String, Value)> {
                    vec!(#(#exprs),*)
                }
            }
        }
        true => quote! {},
    }
}

fn builder(id: &Ident, result: &Type, generics: &Generics) -> proc_macro2::TokenStream {
    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
    let builder_id: syn::Type =
        syn::parse_str(format!("{}Builder", id.to_string()).as_str()).unwrap();
    let builder_func: syn::Expr =
        syn::parse_str(format!("{}Builder::default()", id.to_string()).as_str()).unwrap();

    quote! {
        impl #impl_generics #id #ty_generics #where_clause {
            pub fn builder() -> #builder_id #ty_generics {
                #builder_func
            }
        }

        impl #impl_generics #builder_id #ty_generics #where_clause {
            pub fn exec<C: Client>(
                &self,
                client: &C,
            ) -> Result<Option<#result>, ClientError> {
                self.build().map_err(|e| { ClientError::EndpointBuildError { source: Box::new(e)}})?.exec(client)
            }

            pub fn exec_mut<C: Client, M: MiddleWare>(
                &self,
                client: &C,
                middle: &M,
            ) -> Result<Option<#result>, ClientError> {
                self.build().map_err(|e| { ClientError::EndpointBuildError { source: Box::new(e)}})?.exec_mut(client, middle)
            }
        }
    }
}

fn parse_params(attr: &Meta) -> Result<Parameters, Error> {
    // Parse the attribute as a key/value pair list
    let kv = parse::attr_kv(attr)?;

    // Create map from key/value pair list
    let map = parse::to_map(&kv)?;

    // Convert map to Parameters
    params::Parameters::new(map)
}

fn endpoint_derive(s: synstructure::Structure) -> proc_macro2::TokenStream {
    // Parse `endpoint` attributes attached to input struct
    let attrs = match parse::attributes(&s.ast().attrs) {
        Ok(v) => v,
        Err(e) => return e.into_tokens(),
    };

    // Parse `endpoint` attributes attached to input struct fields
    let field_attrs = match parse::field_attributes(&s.ast().data) {
        Ok(v) => v,
        Err(e) => return e.into_tokens(),
    };

    // Verify attribute is present
    if attrs.is_empty() {
        return Error::new(
            Span::call_site(),
            format!(
                "Deriving `{}` requires attaching an `{}` attribute",
                MACRO_NAME, ATTR_NAME
            )
            .as_str(),
        )
        .into_tokens();
    }

    // Verify there's only one instance of the attribute present
    if attrs.len() > 1 {
        return Error::new(
            Span::call_site(),
            format!("Cannot define the {} attribute more than once", ATTR_NAME).as_str(),
        )
        .into_tokens();
    }

    // Parse endpoint attribute parameters
    let params = match parse_params(&attrs[0]) {
        Ok(v) => v,
        Err(e) => return e.into_tokens(),
    };

    let path = params.path;
    let method = params.method;
    let result = params.result;
    let request_type = params.request_type;
    let response_type = params.response_type;
    let id = &s.ast().ident;

    // Generate action string
    let action = match action(&path) {
        Ok(a) => a,
        Err(e) => return e.into_tokens(),
    };

    // Generate query function
    let query = query(&field_attrs);

    // Gather data
    //gather_attributes(&s.ast().data);

    // Generate helper functions when deriving Builder
    let builder = match params.builder {
        true => builder(&s.ast().ident, &result, &s.ast().generics),
        false => quote! {},
    };

    // Capture generic information
    let (impl_generics, ty_generics, where_clause) = s.ast().generics.split_for_impl();

    // Generate Endpoint implementation
    let const_name = format!("_DERIVE_Endpoint_FOR_{}", id.to_string());
    let const_ident = Ident::new(const_name.as_str(), Span::call_site());
    quote! {
        const #const_ident: () = {
            use ::rustify::client::Client;
            use ::rustify::endpoint::{Endpoint, MiddleWare};
            use ::rustify::enums::{RequestMethod, RequestType, ResponseType};
            use ::rustify::errors::ClientError;
            use ::serde_json::Value;

            impl #impl_generics Endpoint for #id #ty_generics #where_clause {
                type Result = #result;
                const REQUEST_BODY_TYPE: RequestType = RequestType::#request_type;
                const RESPONSE_BODY_TYPE: ResponseType = ResponseType::#response_type;

                fn action(&self) -> String {
                    #action
                }

                fn method(&self) -> RequestMethod {
                    RequestMethod::#method
                }

                #query
            }

            #builder
        };
    }
}

synstructure::decl_derive!([Endpoint, attributes(endpoint)] => endpoint_derive);
