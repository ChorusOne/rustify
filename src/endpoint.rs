use crate::{client::Request, enums::RequestMethod, errors::ClientError};
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use serde_json::Value;
use std::fmt::Debug;
use url::Url;

/// Represents a remote HTTP endpoint which can be executed using a
/// [crate::client::Client].
///
/// This trait can be implemented directly, however, users should prefer using
/// the provided `rustify_derive` macro for generating implementations. An
/// Endpoint consists of:
///   * An `action` which is combined with the base URL of a Client to form a
///     fully qualified URL.
///   * A `method` of type [RequestType] which determines the HTTP method used
///     when a Client executes this endpoint.
///   * A `ResponseType` type which determines the type of response this
///     Endpoint will return when executed.
///
/// Presently, this trait only supports sending and receiving data using JSON.
/// The struct implementing this trait must also implement [serde::Serialize].
/// The fields of the struct act as a representation of data that will be
/// serialized and sent to the remote server. Fields that should be excluded
/// from this behavior can be tagged with the `#[serde(skip)]` attribute. The
/// Endpoint will take the raw response body from the remote server and attempt
/// to deserialize it into the given `ResponseType` which must implement
/// [serde::Deserialize]. This deserialized value is then returned after
/// execution completes.
///
/// Implementations can override the default [transform][Endpoint::transform] in
/// order to modify the raw response content from the remote server before
/// returning it. This is often useful when the remote API wraps all responses
/// in a common format and the desire is to remove the wrapper before returning
/// the deserialized response. It can also be used to check for any errors
/// generated by the API and escalate them accordingly.
///
/// # Example
/// ```
/// use rustify::clients::reqwest::ReqwestClient;
/// use rustify::endpoint::Endpoint;
/// use rustify_derive::Endpoint;
/// use serde::Serialize;
///
/// #[derive(Debug, Endpoint, Serialize)]
/// #[endpoint(path = "my/endpoint")]
/// struct MyEndpoint {}
///
/// // Configure a client with a base URL of http://myapi.com
/// let client = ReqwestClient::default("http://myapi.com");
///     
/// // Construct a new instance of our Endpoint
/// let endpoint = MyEndpoint {};
///
/// // Execute our Endpoint using the client
/// // This sends a GET request to http://myapi.com/my/endpoint
/// // It assumes an empty response
/// let result = endpoint.execute(&client);
/// ```
pub trait Endpoint: Debug + Serialize + Sized {
    /// The type that the raw response from executing this endpoint will
    /// automatically be deserialized to. This type must implement
    /// [serde::Deserialize].
    type Result: DeserializeOwned;

    /// The relative URL path that represents the location of this Endpoint.
    /// This is combined with the base URL from a
    /// [Client][crate::client::Client] instance to create the fully qualified
    /// URL.
    fn action(&self) -> String;

    /// The HTTP method to be used when executing this Endpoint.
    fn method(&self) -> RequestMethod;

    /// Optional query parameters to add to the request
    fn query(&self) -> Vec<(String, Value)> {
        Vec::new()
    }

    /// Combines the given base URL with the relative URL path from this
    /// Endpoint to create a fully qualified URL.
    fn build_url(&self, base: &str) -> Result<url::Url, ClientError> {
        log::info!(
            "Building endpoint url from {} base URL and {} action",
            base,
            self.action()
        );

        let mut url = Url::parse(base).map_err(|e| ClientError::UrlParseError {
            url: base.to_string(),
            source: e,
        })?;
        url.path_segments_mut()
            .unwrap()
            .extend(self.action().split('/'));
        Ok(url)
    }

    /// Executes the Endpoint using the given [Client][crate::client::Client]
    /// and returns the deserialized response as defined by
    /// [Endpoint::Response].
    fn execute<C: crate::client::Client>(
        &self,
        client: &C,
    ) -> Result<Option<Self::Result>, ClientError> {
        log::info!("Executing endpoint");
        log::debug! {"Endpoint: {:#?}", self};

        let url = self.build_url(client.base())?;
        let method = self.method();
        let query = self.query();
        let data = serde_json::to_string(self).map_err(|e| ClientError::DataParseError {
            source: Box::new(e),
        })?;
        let data = match data.as_str() {
            "null" => "".to_string(),
            "{}" => "".to_string(),
            _ => data,
        }
        .into_bytes();
        self.parse(client.execute(Request {
            url,
            method,
            query,
            data,
        }))
    }

    /// Parses the raw response from executing the endpoint into a response type
    /// as defined by [Endpoint::Response].
    fn parse(
        &self,
        res: Result<Vec<u8>, ClientError>,
    ) -> Result<Option<Self::Result>, ClientError> {
        match res {
            Ok(r) => {
                let r_err = r.clone();
                let c = String::from_utf8(r).map_err(|e| ClientError::ResponseConversionError {
                    source: Box::new(e),
                    content: r_err,
                })?;

                log::info!("Parsing JSON result from string");
                log::debug!("Content before transform: {}", c);
                let c = self.transform(c)?;
                log::debug!("Content after transform: {}", c);
                match c.is_empty() {
                    false => Ok(Some(serde_json::from_str(c.as_str()).map_err(|e| {
                        ClientError::ResponseParseError {
                            source: Box::new(e),
                            content: c.clone(),
                        }
                    })?)),
                    true => Ok(None),
                }
            }
            Err(e) => Err(e),
        }
    }

    /// Can be overriden by implementations in order to operate on the raw
    /// response from executing this Endpoint prior to returning the final
    /// response type.
    fn transform(&self, res: String) -> Result<String, ClientError> {
        Ok(res)
    }
}

/// Represents an empty Endpoint result.
#[derive(Deserialize, Debug)]
pub struct EmptyEndpointResult {}
