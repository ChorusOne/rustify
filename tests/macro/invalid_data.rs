use rustify::endpoint::Endpoint;
use rustify_derive::Endpoint;
use serde::Serialize;

#[derive(Debug, Endpoint, Serialize)]
#[endpoint(path = "test/path")]
struct Test {
    pub name: String,
    #[endpoint(data)]
    pub data: String,
}

#[derive(Debug, Endpoint, Serialize)]
#[endpoint(path = "test/path")]
struct TestTwo {
    pub name: String,
    #[endpoint(data)]
    pub data: Vec<u8>,
    #[endpoint(data)]
    pub data_two: Vec<u8>,
}

fn main() {}
