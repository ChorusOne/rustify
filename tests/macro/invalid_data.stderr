error: May only mark one field as the data field
  --> $DIR/invalid_data.rs:20:9
   |
20 |     pub data_two: Vec<u8>,
   |         ^^^^^^^^

error[E0308]: mismatched types
 --> $DIR/invalid_data.rs:5:17
  |
5 | #[derive(Debug, Endpoint, Serialize)]
  |                 ^^^^^^^^ expected slice `[u8]`, found struct `std::string::String`
  |
  = note: expected reference `&[u8]`
             found reference `&std::string::String`
  = note: this error originates in the derive macro `Endpoint` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unused import: `rustify::endpoint::Endpoint`
 --> $DIR/invalid_data.rs:1:5
  |
1 | use rustify::endpoint::Endpoint;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
